common/os.go

```diff
@@ -124,32 +124,32 @@ func MustWriteFile(filePath string, contents []byte, mode os.FileMode) {
 	}
 }
 
-// WriteFileAtomic writes newBytes to temp and atomically moves to filePath
-// when everything else succeeds.
-func WriteFileAtomic(filePath string, newBytes []byte, mode os.FileMode) error {
-	dir := filepath.Dir(filePath)
-	f, err := ioutil.TempFile(dir, "")
+// WriteFileAtomic creates a temporary file with data and the perm given and
+// swaps it atomically with filename if successful.
+func WriteFileAtomic(filename string, data []byte, perm os.FileMode) error {
+	var (
+		dir      = filepath.Dir(filename)
+		tempFile = filepath.Join(dir, "write-file-atomic-"+RandStr(32))
```

Comment:
Am a little hesitant on this new approach, the previous `ioutil.TempFile` handled that already with all the permissions and wouldn't add the tempfile to the same working directory as this.

Reply:
The old implementation used the directory of the target file, this is why issues like tendermint/tendermint#1329 surface. Sticking to the current approach might just be the best as we need to expect proper rights to be able to write to that directory, any other target directory might be too much of an assumption on the environment.

Reply:
>The old implementation used the directory of the target file, this is why issues like 

Thank you for the correction @xla, for some reason I assumed that since ioutil.TempDir adds them to the tempdir, vicariously ioutil.Tempfile did too, but I should have known better ;) However, your approach adds the file to be renamed to the final destination's directory too with `filepath.Join(parentDir, generatedSuffix)`, right?
I am just asking to understand the change better.

Reply:
`ioutil.TempDir` only uses `os.Tempdir` if the dir argument is left blank, same for `ioutil.TempFile`.

> However, your approach adds the file to be renamed to the final destination's directory too with filepath.Join(parentDir, generatedSuffix), right?

Indeed, while it is tempting to resort to the platform specific temp directories, it lacks guarantees, from the docs:

> The directory is neither guaranteed to exist nor have accessible permissions.